A4A > Marketplace: Implement term pricing for Hosting, Products & Cart

yashwin opened the pull request:
Part of https://linear.app/a8c/issue/A4A-1926/marketplace-show-monthlyearly-pricing-for-hosting-and-products

Resolves https://linear.app/a8c/issue/A4A-1929/products-display-monthlyyearly-pricing-on-cards
Resolves https://linear.app/a8c/issue/A4A-1930/hosting-display-monthlyyearly-pricing-on-cards
Resolves https://linear.app/a8c/issue/A4A-1961/cart-display-monthlyyearly-pricing
Resolves https://linear.app/a8c/issue/A4A-1962/display-term-based-pricing-on-product-details-modal-lightbox

## Proposed Changes

This PR implements term pricing for Hosting, Products & Cart.

## Why are these changes being made?

* To support term pricing on A4A Marketplace

## Testing Instructions

Note: There might be some pricing difference when the feature flag is enabled. Please ignore it now, as we will fix it later. 

* Open the A4A live link
* Use a non-BD agency > Test all the scenarios mentioned below > Verify the prices match production, and all the flows work as expected.
* Use a BD agency > Append the URL with `?flags=-a4a-bd-term-pricing`
* Test all the scenarios mentioned below > Verify the prices match production, and all the flows work as expected.

1) Marketplace > Hosting > Standard Agency Hosting > Refer mode on/off
2) Marketplace > Hosting > Premier Agency Hosting > Refer mode on/off
3) Marketplace > Products > Refer mode on/off
4) Marketplace > Products > View details for a product
5) Marketplace > Checkout

* Verify that all the filters work as expected on the Products page

* Patch 199438-ghe-Automattic/wpcom/
* Append the URL with `?flags=a4a-bd-term-pricing` > Test all the scenarios mentioned below > Verify the UI looks as displayed below for yearly and monthly billing. 

Billing term: Monthly

1) Marketplace > Hosting > Standard Agency Hosting > Refer mode off

<img width="1920" height="860" alt="Screenshot 2026-01-14 at 2 48 42 PM" src="https://github.com/user-attachments/assets/c250e1b6-4e42-42e2-a95e-3759c6720581" />

2) Marketplace > Hosting > Standard Agency Hosting > Refer mode on

<img width="1920" height="860" alt="Screenshot 2026-01-14 at 2 53 55 PM" src="https://github.com/user-attachments/assets/d0e88cf9-66b0-43bd-b9ed-0271b300083f" />

3) Marketplace > Hosting > Premier Agency Hosting > Refer mode on

<img width="1920" height="860" alt="Screenshot 2026-01-14 at 2 54 31 PM" src="https://github.com/user-attachments/assets/723526a7-3897-4ee8-8339-d2e87c988d8e" />

4) Marketplace > Hosting > Premier Agency Hosting > Refer mode off

<img width="1920" height="860" alt="Screenshot 2026-01-14 at 2 55 27 PM" src="https://github.com/user-attachments/assets/b38a12be-0310-44e0-9242-a81dbb30937c" />

5) Marketplace > Products > Refer mode off 

<img width="1920" height="860" alt="Screenshot 2026-01-14 at 2 57 19 PM" src="https://github.com/user-attachments/assets/307fef64-78c4-4e9a-96e9-736076e0020d" />

6) Marketplace > Products > Refer mode on 

<img width="1920" height="860" alt="Screenshot 2026-01-14 at 2 57 54 PM" src="https://github.com/user-attachments/assets/8e35deec-3dd5-4d89-aaa7-ae0019039a7b" />

7) Cart > Refer mode off

<img width="663" height="860" alt="Screenshot 2026-01-14 at 2 58 54 PM" src="https://github.com/user-attachments/assets/6fadfebc-ff46-4766-a8f0-241142d65eae" />

8) Cart > Refer mode on 

<img width="663" height="860" alt="Screenshot 2026-01-14 at 2 59 06 PM" src="https://github.com/user-attachments/assets/1e62dd55-539e-4fae-ae2b-3d2538bb8a14" /> 

9) View details 

<img width="1125" height="641" alt="Screenshot 2026-01-14 at 3 04 54 PM" src="https://github.com/user-attachments/assets/49f5b3be-03d6-4e87-8bc0-3bf14fa1aa7d" />


Billing term: Yearly


1) Marketplace > Hosting > Standard Agency Hosting > Refer mode off

<img width="1920" height="881" alt="Screenshot 2026-01-14 at 3 00 53 PM" src="https://github.com/user-attachments/assets/d6c3a130-eeba-4889-82b0-2ae5470ad7ff" />

2) Marketplace > Hosting > Standard Agency Hosting > Refer mode on

<img width="1920" height="881" alt="Screenshot 2026-01-14 at 3 01 04 PM" src="https://github.com/user-attachments/assets/495859fe-a22e-41f3-ac61-0c30c79efd49" />

3) Marketplace > Hosting > Premier Agency Hosting > Refer mode off

<img width="1920" height="881" alt="Screenshot 2026-01-14 at 3 01 27 PM" src="https://github.com/user-attachments/assets/ac1790d0-1d44-4566-b9f1-bcde854adaca" />

4) Marketplace > Hosting > Premier Agency Hosting > Refer mode on

<img width="1920" height="881" alt="Screenshot 2026-01-14 at 3 02 00 PM" src="https://github.com/user-attachments/assets/9ecd26fa-3b4c-4e0f-ab0d-427ed9c43cd9" />

5) Marketplace > Products > Refer mode off 

<img width="1920" height="881" alt="Screenshot 2026-01-14 at 3 03 04 PM" src="https://github.com/user-attachments/assets/7f75ef90-f48b-4acb-9cde-afe5ed4492f4" />

6) Marketplace > Products > Refer mode on 

<img width="1920" height="881" alt="Screenshot 2026-01-14 at 3 03 16 PM" src="https://github.com/user-attachments/assets/35cd0ea1-161e-4a32-a2f1-b82a1568d5a6" />

7) Cart > Refer mode off

<img width="647" height="881" alt="Screenshot 2026-01-14 at 3 03 30 PM" src="https://github.com/user-attachments/assets/9c3c8394-e02e-4642-861b-13020793344f" />

8) Cart > Refer mode on 

<img width="647" height="881" alt="Screenshot 2026-01-14 at 3 03 41 PM" src="https://github.com/user-attachments/assets/0799b283-83d5-4020-bcda-729b0e8109b3" />

9) View details

<img width="1099" height="644" alt="Screenshot 2026-01-14 at 3 04 43 PM" src="https://github.com/user-attachments/assets/e69758a7-a7d5-4f12-b610-cc586e6b1c6d" />

## Pre-merge Checklist



<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
A4A > Marketplace: Implement Term pricing for Hosting, Products & Cart

jkguidaven commented on client/a8c-for-agencies/sections/marketplace/shopping-cart/shopping-cart-menu/item.tsx:
```diff
@@ -32,31 +36,49 @@ export default function ShoppingCartMenuItem( { item, onRemoveItem }: ItemProps
 	const isFree = actualCost === 0;
 
+	const renderPrice = () => {
+		if ( isTermPricingEnabled ) {
```
I can see that we are doing conditional rendering with the prices based on the term selected. I wonder if we could just handle the calculation from `getProductPricingInfo` so we don't have to do conditional rendering here and other places. wdyt?

yashwin commented on client/a8c-for-agencies/components/pressable-usage-details/index.tsx:
```diff
@@ -6,7 +6,7 @@ import { useSelector } from 'react-redux';
 import { getActiveAgency } from 'calypso/state/a8c-for-agencies/agency/selectors';
 import { TitanOrder } from 'calypso/state/a8c-for-agencies/types';
-import { APIProductFamilyProduct } from 'calypso/state/partner-portal/types';
+import type { APIProductFamilyProduct } from 'calypso/a8c-for-agencies/types/products';
```
We created a new types file on A4A as we are introducing new fields.

yashwin commented on client/a8c-for-agencies/data/marketplace/use-products-query.ts:
```diff
@@ -60,20 +46,9 @@ function queryProducts(
 		} );
 }
 
-export function usePublicProductsQuery(): UseQueryResult< APIProductFamilyProduct[], unknown > {
```
We are not using this anywhere.

yashwin commented on client/a8c-for-agencies/data/marketplace/use-products-query.ts:
```diff
@@ -60,20 +46,9 @@ function queryProducts(
+const getProductsQueryKey = ( agencyId?: number ) => [ 'a4a', 'marketplace', 'products', agencyId ];
 
 export default function useProductsQuery(
-	isPublicFacing = false,
```
Removed support for the public-facing API as we no longer need it.

yashwin commented on client/a8c-for-agencies/sections/marketplace/hooks/use-product-and-plans.tsx:
```diff
@@ -111,9 +111,8 @@ export default function useProductAndPlans( {
 	productSearchQuery,
-	usePublicQuery = false,
 }: Props ) {
-	const { data, isLoading: isLoadingProducts } = useProductsQuery( usePublicQuery );
```
usePublicQuery is not used anywhere.

yashwin commented on client/a8c-for-agencies/sections/marketplace/hooks/use-product-categories.ts:
```diff
@@ -24,7 +24,7 @@ type CategoryConfig = {
 export function useProductCategories( product: APIProductFamilyProduct ): string[] {
 	const translate = useTranslate();
-	const { family_slug } = product;
+	const { family_slug, slug } = product;
```
We are going to change how we group families. You can see the response here: 199438-ghe-Automattic/wpcom

To support that, we are going to use the slug also so that the current implementation doesn't break.

yashwin commented on client/a8c-for-agencies/sections/marketplace/hooks/use-total-invoice-value.ts:
It's ok to repeat the same return here as we will get rid of the logic completely when we enable BD for all the users.

yashwin commented on client/a8c-for-agencies/sections/marketplace/products-overview/hocs/with-product-lightbox.tsx:
```diff
@@ -138,16 +151,23 @@ function withProductLightbox< T >(
 			</>
 		);
 	};
+
```
Fixing some TS warnings

yashwin commented on client/a8c-for-agencies/sections/referrals/lib/commissions.ts:
```diff
@@ -24,29 +21,15 @@ export const isProductEligibleForCommission = ( slug: string ) => {
 	return 0;
 };
-
-export const calculateCommissions = ( referral: Referral, products: APIProductFamilyProduct[] ) => {
```
Not being used anymore

yashwin commented on client/a8c-for-agencies/sections/referrals/lib/commissions.ts:
```diff
@@ -24,29 +21,15 @@ export const isProductEligibleForCommission = ( slug: string ) => {
 
-export const getProductCommissionPercentage = ( slug?: string ) => {
-	if ( ! slug || ! isProductEligibleForCommission( slug ) ) {
+export const getProductCommissionPercentage = ( slug?: string, familySlug?: string ) => {
```
As mentioned earlier, we will change the family structure. This is to correctly calculate the commissions.

cleacos replied:
yes, both are needed

jkguidaven commented on client/a8c-for-agencies/sections/marketplace/hosting-overview/hosting-content/premier-agency-hosting/pressable-plan-section/regular-plan-card-content.tsx:
```diff
@@ -48,6 +52,21 @@ export default function PressablePlanSelectorCard( {
 
+	const isTermPricingEnabled = isEnabled( 'a4a-bd-term-pricing' ) && isEnabled( 'a4a-bd-checkout' );
+
+	const priceInterval = () => {
```
Is there a way we could make this reusable? This seems to be repeated 3 times in different places. could be part of `getProductPricingInfo`?

yashwin replied:
I'd tried doing it. The thing is, this will be removed soon, so we will only have term pricing going forward. So, I would leave this as-is for now. Also, it is repeated in 2 places only, the text is different in another place.

yashwin committed:
fix filters

yashwin commented:
Thanks for the review, @cleacos & @jkguidaven!

I found the filters weren't working properly. I have fixed it now: https://github.com/Automattic/wp-calypso/pull/107867/commits/31110d83a1c446551dd6479a03a87eb7e34874c8

> I tested it, using the new endpoint, it seems that it works well, but my checkout is configured to use EUR, so in the marketplace the prices are in USD, and in the checkout it is EUR.

@cleacos This is based on the endpoint, not the UI. UI is configured to work with the product's currency. I have now updated the endpoint to take the current based on the agency. So, this should work fine now.

yashwin committed:
fix tests

cleacos commented:
The "Jetpack VaultPress Backup Add-ons" are displayed as "Free" in the yearly mode. 

<img width="868" height="653" alt="image" src="https://github.com/user-attachments/assets/c7bf321f-e18a-4109-95ce-d905431baedf" />

We should hide/filter out product items with null pricing.

yashwin committed:
minor fixes

yashwin committed:
fixes and disable yearly term for referrals

yashwin commented:
> But... the checkout doesn't work (both: monthly and yearly plans). Maybe because of the product ID.

@cleacos Fixed it. Please take the latest backend code.

yashwin committed:
fix alternative_product_id

cleacos commented on client/a8c-for-agencies/types/products.ts:
```diff
@@ -0,0 +1,40 @@
+	product_id: number;
+	monthly_product_id?: number;
+	yearly_product_id?: number;
+	alternative_product_id?: number;
```
~~Missed fields?~~

- ~~`monthly_alternative_product_id`~~
- ~~`yearly_alternative_product_id`~~

~~and remove `alternative_product_id`?~~

I had this comment from the previous version

cleacos commented on client/a8c-for-agencies/sections/marketplace/lib/product-slugs.ts:
```diff
@@ -1,9 +1,8 @@
 export const SECURITY_PRODUCT_SLUGS = [
-	'jetpack-backup',
+	'jetpack-backup-t1',
```
I think that Jetpack Backup products could be T1 or T2.

yashwin replied:
No, we only show T1, right?

yashwin commented on client/a8c-for-agencies/sections/marketplace/hosting-overview/hosting-content/standard-agency-hosting/wpcom-plan-section/wpcom-plan-selector.tsx:
```diff
@@ -55,7 +61,10 @@ export default function WPCOMPlanSelector( {
+	const termPrice = termPricing === 'yearly' ? plan.yearly_price ?? 0 : plan.monthly_price ?? 0;
+	const productPrice = isTermPricingEnabled ? termPrice : Number( plan?.amount ?? 0 );
+
+	const originalPrice = Number( productPrice ) * quantity;
```
Yes, it will be always 1
